purchase_order_product_recommendation/wizards/purchase_order_recommendation.py

```diff
@@ -0,0 +1,297 @@
+                ('location_id.usage', '=', src),
+                ('location_dest_id.usage', '=', dst),
+                ('state', '=', 'done'),
+            ],
+            ['product_id', 'qty_done'],
+            ['product_id'],
+        )
+        # Manual ordering that circumvents ORM limitations
+        found_lines = sorted(
+            found_lines,
+            key=lambda res: (
+                res['product_id_count'],
+                res['qty_done'],
+            ),
+            reverse=True,
+        )
+        return {l['product_id'][0]: l for l in found_lines}
+
+    @api.model
+    def _add_product_values(self, product_ids, found_dict):
+        """Can be used to override in order to add product without refetching
+           all the product records again"""
+        for product in product_ids:
+            line = found_dict[product.id]
+            line['qty_available'] = product.qty_available
+            line['qty_virtual_available'] = product.virtual_available
+            if not line.get('qty_delivered'):
+                continue
+            # Simple forecast based on quantities needed to fulfill the demand
+            # of a product in the same period of time.
+            qty_to_order = abs(
+                min(0, line['qty_virtual_available'] - line['qty_delivered']))
+            line['qty_to_order'] = qty_to_order
+            line['is_modified'] = bool(qty_to_order)
+            line['qty_avg'] = line['qty_delivered'] / self._get_total_days()
+        return found_dict
+
+    @api.model
+    def _prepare_wizard_line(self, vals, order_line=False):
+        """Used to create the wizard line"""
+        return {
+            'purchase_line_id': order_line and order_line.id,
+            'product_id': (order_line and order_line.product_id.id or
+                           vals['product_id'][0]),
+            'times_delivered': vals.get('times_delivered', 0),
+            'times_received': vals.get('times_received', 0),
+            'units_received': vals.get('qty_received', 0),
+            'units_available': vals.get('qty_available', 0),
+            'units_virtual_available': vals.get('qty_virtual_available'),
+            'units_avg_delivered': vals.get('qty_avg', 0),
+            'units_delivered': vals.get('qty_delivered', 0),
+            'units_included': (order_line and order_line.product_qty or
+                               vals.get('qty_to_order', 0)),
+            'is_modified': vals.get('is_modified', False),
+        }
+
+    @api.multi
+    @api.onchange('order_id', 'date_begin', 'date_end', 'line_amount')
+    def _generate_recommendations(self):
+        """Generate lines according to received and delivered items"""
+        last_compute = '{}-{}-{}-{}'.format(
+            self.id, self.date_begin, self.date_end, self.line_amount)
+        # Avoid execute onchange as times as fields in api.onchange
+        # ORM must control this?
+        if self.last_compute == last_compute:
```

Comment:
Why is this needed?

Reply:
In `sale_order_product_recommendation` is suposed to avoid running the onchange for every field declared, but indeed a made a debug and that's not happening. Can bring some light to this, @Yajo ?

Reply:
No idea, but indeed it seems redundant with the onchange decorator. Is it possible that this comes from v7, where we didn't have such decorator?

Reply:
Well, this module was first developed for v9, so I don't think so. Let's remove it.